openedx/core/djangoapps/catalog/utils.py

```diff
@@ -114,45 +120,149 @@ def munge_catalog_program(catalog_program):
     }
 
 
-def get_course_run(course_key, user):
-    """Get a course run's data from the course catalog service.
-
-    Arguments:
-        course_key (CourseKey): Course key object identifying the run whose data we want.
-        user (User): The user to authenticate as when making requests to the catalog service.
-
-    Returns:
-        dict, empty if no data could be retrieved.
+def get_course_runs_from_catalog_and_cache_it(user, course_keys):
+    """
+    Get course run's data from the course catalog service and cache it.
     """
+    catalog_course_runs_against_course_keys = {}
     catalog_integration = CatalogIntegration.current()
-
     if catalog_integration.enabled:
-        api = create_catalog_api_client(user, catalog_integration)
+        api = course_discovery_api_client(user)
 
-        data = get_edx_api_data(
+        catalog_data = get_edx_api_data(
             catalog_integration,
             user,
             'course_runs',
-            resource_id=unicode(course_key),
-            cache_key=catalog_integration.CACHE_KEY if catalog_integration.is_cache_enabled else None,
             api=api,
-            querystring={'exclude_utm': 1},
+            querystring={'keys': ",".join(course_keys), 'exclude_utm': 1},
         )
+        if catalog_data:
+            for catalog_course_run in catalog_data:
+                CatalogCourseRunCacheUtility.cache_course_run(catalog_course_run)
+                catalog_course_runs_against_course_keys[catalog_course_run["key"]] = catalog_course_run
+    return catalog_course_runs_against_course_keys
 
-        return data if data else {}
-    else:
-        return {}
 
+def get_course_runs(user, course_keys=None):
+    """
+    Get course run data from the course catalog service if not available in cache.
+
+    Arguments:
+        course_keys ([String]): A list of Course key strings identifying the run whose data we want.
```

Comment:
@rlucioni 
The whole flow of [retreiving user enrollments (with course_id as string in model object)](https://github.com/edx/edx-platform/pull/13235/files/48b4529667372e744da3cdfc622bab5ba211e8d0#diff-60b5dd29d4b2ee6a9f3eae0a569d5210R274) and then [getting catalog course runs for batch of course keys](https://github.com/edx/edx-platform/pull/13235/files/48b4529667372e744da3cdfc622bab5ba211e8d0#diff-d84de42b66532a7a01cdf6ac15d6fb96R137), [caching catalog data against course keys](https://github.com/edx/edx-platform/pull/13235/files/48b4529667372e744da3cdfc622bab5ba211e8d0#diff-d84de42b66532a7a01cdf6ac15d6fb96R250), [retrieving cache against course keys](https://github.com/edx/edx-platform/pull/13235/files/48b4529667372e744da3cdfc622bab5ba211e8d0#diff-d84de42b66532a7a01cdf6ac15d6fb96R239); all of it involves string course key values instead of `CourseKey` objects.


Isn't it better that instead of converting back and forth between `CourseKey` objects and `unicode()` values, we convert it just once and use the same value in the all of the following code flow?